feat: Implement auto bounty payout for merged PRs (fixes #3941)

TAGOOZ opened the pull request:
Fixes #3941

## Summary

This PR implements automatic bounty payment processing when a pull request is merged that closes an issue with a bounty label. The system detects merged PRs, extracts linked issues, verifies bounty labels, and initiates payment through the GitHub Sponsors API.

## Changes

### Database
- Added `preferred_payment_method` field to UserProfile model to store user's payment preference (GitHub Sponsors or Bitcoin Cash)
- Created migration `0247_userprofile_preferred_payment_method.py`

### Backend API
- Created new endpoint `/api/bounty_payout/` in `website/views/organization.py`
- Accepts POST requests with issue URL and optional PR URL
- Validates issue has bounty label and assignee
- Checks if user exists in BLT with linked GitHub account
- Prevents duplicate payments by checking existing transaction IDs
- Returns appropriate error messages for various failure cases
- Added optional authentication via `X-BLT-API-TOKEN` header
- Added URL route in `blt/urls.py`

### GitHub Actions
- Created workflow `.github/workflows/auto-bounty-payout.yml`
- Triggers when PR is merged (not just closed)
- Extracts linked issues from PR body using regex patterns
- Checks each issue for dollar-amount labels ($5, $10, etc.)
- Calls BLT API endpoint with issue details
- Posts comment on issue confirming payment initiation or reporting errors

## Implementation Details

The workflow looks for PR descriptions containing "Fixes #123", "Closes #456", or "Resolves #789" patterns to identify linked issues. For each linked issue with a bounty label, it makes an API call to process the payment.

The API endpoint validates the request, fetches issue details from GitHub, verifies the assignee has a BLT account with connected GitHub profile, checks their payment preference, and records the transaction. Currently uses placeholder transaction IDs pending full GitHub Sponsors GraphQL API integration.

## Improvements Over Previous Attempt (PR #4236)

Based on CodeRabbit's review of the previous implementation, this version addresses several critical issues:
- Uses curl for API calls instead of github.request() which only works for GitHub's own API
- Properly reads authentication headers using request.headers.get() instead of making HTTP requests
- No duplicate class definitions
- Consistent data types throughout
- Correct migration dependencies
- Added duplicate payment prevention

## Known Limitations

- GitHub Sponsors API integration uses placeholder transaction IDs. Full GraphQL API implementation will be added in future work.
- Only GitHub Sponsors payment method is currently implemented. Bitcoin Cash support is planned.
- Requires BLT_API_TOKEN to be configured in settings and GitHub repository secrets.
- Workflow API URL needs to be updated to production URL before deployment.

## Testing

All Python files compile without syntax errors. Migration dependency has been verified against latest migration (0246). Error handling covers all expected failure cases including missing assignee, invalid issue URL, user not found, and duplicate payments.

Manual testing required after deployment to verify end-to-end workflow.

## Deployment Notes

1. Run migration: `python manage.py migrate`
2. Add BLT_API_TOKEN to Django settings
3. Add BLT_API_TOKEN to GitHub repository secrets
4. Update API URL in workflow file (line 74) to production URL
5. Restart application server

This is my first contribution to the project. I've tried to follow the existing code patterns and address feedback from previous attempts. Happy to make any changes based on review feedback.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Automated bounty payout processing when merged PRs reference issues
  * User preference for bounty payment method added
  * Sponsorship payout tracking and cancellation/retry improvements

* **Documentation**
  * Setup docs and example env variables for bounty payout automation

* **Tests / CI**
  * Improved test reliability with serialized DB handling and cleanup

* **Chores**
  * Ignore test DB file in .gitignore
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
feat: Implement auto bounty payout for merged PRs (fixes #3941)

- Add preferred_payment_method field to UserProfile model
- Create /api/bounty_payout/ endpoint with authentication
- Add GitHub Action workflow to trigger on PR merge
- Implement duplicate payment prevention
- Add comprehensive error handling and validation

Addresses issues found in CodeRabbit review of PR #4236:
- Uses curl instead of github.request() for API calls
- Proper header authentication (X-BLT-API-TOKEN)
- No duplicate class definitions
- Consistent data types
- Fixed migration dependencies

Commit:
fix: address CodeRabbit and security review feedback

- Add explicit permissions block to workflow (least privilege)
- Fix script injection vulnerability by using environment variable for PR body
- Make API URL configurable via BLT_API_URL secret
- Fix exception handling to not expose stack traces
- Fix linting issues (unused variable, f-string without placeholders)
- Use logger.exception instead of logger.error for better debugging

DonnieBLT commented on website/views/organization.py:
```diff
@@ -3742,3 +3742,212 @@ def extract_payment_info_from_comments(self, issue, owner, repo_name, issue_numb
+            }
+        )
+        
+        # Mark as paid with sponsors (placeholder transaction ID)
```
Please add the logic to process the payment

Commit:
feat: implement GitHub Sponsors API integration and address all review feedback

Addresses feedback from:
- @DonnieBLT: 'Please add the logic to process the payment'
- CodeRabbit: URL validation, unused variables, import order, error messages
- Copilot AI: Import placement, UUID for transaction IDs, error specificity

GitHub Sponsors Integration:
- Fetch user's sponsorable ID and available tiers via GraphQL
- Find matching tier for bounty amount (or closest available)
- Create sponsorship via createSponsorship mutation
- Store real sponsorship ID from GitHub API
- Handle GitHub's broken one-time payment API (uses tier-based instead)

Security & Code Quality Fixes:
- Move duplicate payment check BEFORE API call (prevents double charges)
- Add URL validation with regex for owner/repo/issue_number
- Move imports to top of file (SocialAccount, uuid)
- Make sponsor username configurable via GITHUB_SPONSOR_USERNAME setting
- Use UUID instead of timestamp for better uniqueness
- Improve error messages with specific details
- Remove unused exception variable
- Fix import order for isort compliance
- Add comprehensive error handling for all scenarios

Technical Details:
- Uses GitHub GraphQL API (not REST)
- Validates user has GitHub Sponsors enabled
- Checks for existing tiers before payment
- Handles network errors and API failures gracefully
- Logs all operations for debugging

Requirements:
- GITHUB_TOKEN with user/read:org scopes
- Recipients must have GitHub Sponsors with tiers configured
- GITHUB_SPONSOR_USERNAME setting (defaults to DonnieBLT)
- BLT_API_TOKEN for webhook authentication (optional)

Known Limitations:
- GitHub one-time payment API is broken (documented issue)
- Uses recurring sponsorships (can be cancelled manually)
- Requires matching or closest tier amount
- No sandbox mode (creates real sponsorships)

Commit:
fix: correct import order for isort (allauth before bs4)

DonnieBLT commented on PR_DESCRIPTION.md:
Delete

TAGOOZ committed:
Delete PR_DESCRIPTION.md

Commit:
feat: safe sponsors payout with cancel and validation

Commit:
- Add CSRF exemption for GitHub Actions webhook calls
- Verify PR merge status before processing bounty payouts
- Improve error logging with logger.exception() in exception handlers
- Fix EN DASH character in bounty amount comment

Addresses security issues identified in CodeRabbit review:
- Prevents CSRF failures for automated GitHub Actions calls
- Ensures only merged PRs trigger bounty payments
- Provides better error context with stack traces

Commit:
- Add detailed error messages for different HTTP status codes (404, 401/403, 429)
  to provide clearer troubleshooting guidance for users
- Replace generic error message with actionable feedback for common failure scenarios
- Improves user experience when GitHub API calls fail during bounty payout processing

Commit:
feat: add sponsors cancellation tracking and improve error handling

- Add GitHubIssue fields for sponsors cancellation observability
- Expose failure tracking in admin for quick triage
- Clean up error messages to avoid exposing exception details
- Set/clear cancellation flags in process_bounty_payout

Commit:
fix: require exact tier match for bounty payments

- Remove tier rounding up logic
- Only proceed with payment if exact tier exists
- Return clear error if no matching tier found

Commit:
security: add repository allowlist for bounty payouts

CRITICAL: Prevent unauthorized repos from draining sponsors budget
- Add BLT_ALLOWED_BOUNTY_REPOS setting (default: OWASP-BLT/BLT)
- Reject payout requests for repos not in allowlist (403)
- Log unauthorized attempts for security monitoring
- Add env-based configuration for multiple repos

Without this, attackers could create fake issues in their own repos
with high dollar labels and drain the entire sponsors budget.

Commit:
feat: add bounty payout configuration to environment setup

Commit:
Fix SQLite savepoint error in Selenium tests

- Switch test database to file-based SQLite to avoid transaction conflicts
- Add @tag('serial') to MySeleniumTests to prevent parallel execution
- Add serialized_rollback=True for proper transaction handling
- Update .gitignore to exclude test_db.sqlite3
- Add cleanup step in CI/CD workflow

Fixes: sqlite3.OperationalError: cannot open savepoint - SQL statements in progress

DonnieBLT reviewed (changes requested):
looks great can you please resolve the conflicts

Commit:
Merge remote-tracking branch 'origin/main' into feat/auto-bounty-payout

Commit:
fix: address CodeRabbit security and reliability concerns

- Enforce BLT_API_TOKEN validation (fail-closed if not configured)
- Validate PR-issue linkage via GitHub timeline API
- Verify PR and issue are from same repository
- Block retry attempts after failed cancellation
- Validate GITHUB_TOKEN before GraphQL calls
- Add ClassVar type hint for serialized_rollback
